### ISSUE TITLE
CpGs towards the start and end of a read not considered by Nanopolish?

### ISSUE THREAD

**Issue by lkerr34** [jts/nanopolish#1058]
Hi,

I have been using Nanopolish to call methylation and wondered if reads are automatically trimmed by Nanopolish when making these methylation calls? My reason for asking is that when I compare read alignments from Minimap2 to the methylation tsv calls I see that some CpGs towards the start and end of a read appear to be missing from the tsv file.

I'd really appreciate any info you could give me!

Thanks,
Lyndsay

**Comment by jts** [comment]
Yes, there is internal trimming. The signal-to-read alignment step can discard the ends of the read, and also the methylation calling algorithm avoids calling at the ends of the signal boundaries. There may be other steps as well but those are the two off the top of my head.

**Comment by lkerr34** [comment]
That's great - thank you so much for the quick reply! 

I guess that the end of signal boundaries are the read ends? I have found that I also have a few examples where CpGs further into a read are missing from the tsv file (even when surrounding CpGs are included). Can you think of anything that may cause this?

**Comment by jts** [comment]
Nanopolish skips regions that have long regions of high CpG density as it is computationally too expensive to call this - is this possible for this case?

**Comment by lkerr34** [comment]
I don't think this can be the case here as when I run the calculate methylation frequency script some of the CpGs that seem to be missed on individual reads appear (and I assume Nanopolish would exclude these high-density regions from all reads so that they wouldn't appear in the bulk data).

**Comment by lkerr34** [comment]
I've attached a specific example of a short read where the discrepancy occurs in case this is useful.

The first screenshot is the entries in the Nanopolish tsv files associated with a particular read. 6 CpGs are included. 

![meth_tsv](https://user-images.githubusercontent.com/55538405/218995887-3fe7bd3d-6b3f-4472-aee3-abed481fd730.png)

The next screenshot is the corresponding Minimap2 alignment for the read. Here, 7 CpGs are detected (highlighted), with the CpG highlighted in green being the one that is missing in the Nanopolish tsv file.

![minimap2_alignment](https://user-images.githubusercontent.com/55538405/218996151-b4e05702-c42f-4c99-89a9-ccb6dda5a188.png)

Finally, this is a snippet from the calculate methylation frequency script, where all 7 of the CpGs in the region are included.

![bulk_meth](https://user-images.githubusercontent.com/55538405/218997029-0750f742-40ba-4d49-9c1d-5a3ff2bfe470.png)

So this CpG is being omitted from this read but not from other reads.

**Comment by jts** [comment]
Hm, it might be this check then:

https://github.com/jts/nanopolish/blob/21c75db77b0e1338989be2d5c6bc96eca7c86c37/src/basemods/nanopolish_basemods.cpp#L354

You could try commenting it out to see if the call at 307 appears.

**Comment by lkerr34** [comment]
Apologies if this is a silly question but I installed Nanopolish via conda and can't find the nanopolish_basemods.cpp file you mention. Would you expect me to have access to that file if I installed via conda?

Also, can I ask what it is that this part of the script checks? What are the boundaries?
